File: Assignment7/Stats.py
class StatisticsCalculator:
    def __init__(self, data):
        self.data = data

    def mean(self):
        return sum(self.data) / len(self.data)

    def median(self):
        sorted_data = sorted(self.data)
        n = len(sorted_data)
        mid = n // 2
        if n % 2 == 0:
            return (sorted_data[mid - 1] + sorted_data[mid]) / 2
        else:
            return sorted_data[mid]

    def mode(self):
        frequency = {}
        for num in self.data:
            frequency[num] = frequency.get(num, 0) + 1
        max_count = max(frequency.values())
        modes = [k for k, v in frequency.items() if v == max_count]
        return modes, max_count

if __name__ == "__main__":
    data = [1, 2, 2, 3, 4]
    calc = StatisticsCalculator(data)

    print(f"Mean: {calc.mean():.2f}")
    print(f"Median: {calc.median():.2f}")
    modes, count = calc.mode()
    print(f"Mode(s): {modes} (occurs {count} times)")

File: Assignment7/Stats.ml
let mean lst =
  let sum = List.fold_left (+) 0 lst in
  float_of_int sum /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let len = List.length sorted in
  if len mod 2 = 0 then
    let mid1 = List.nth sorted (len / 2 - 1)
    and mid2 = List.nth sorted (len / 2) in
    (float_of_int (mid1 + mid2)) /. 2.0
  else
    float_of_int (List.nth sorted (len / 2))

let mode lst =
  let freq_map = List.fold_left (fun acc x ->
    if List.mem_assoc x acc then
      let count = List.assoc x acc + 1 in
      (x, count) :: List.remove_assoc x acc
    else
      (x,1)::acc
  ) [] lst in
  let (mode_val, mode_count) = List.fold_left (fun (m_val,m_cnt) (k,v) ->
    if v > m_cnt then (k,v) else (m_val,m_cnt)
  ) (0,0) freq_map in
  (mode_val, mode_count)

let () =
  let data = [1;2;2;3;4] in
  Printf.printf "Mean: %.2f\n" (mean data);
  Printf.printf "Median: %.2f\n" (median data);
  let (m_val, m_count) = mode data in
  Printf.printf "Mode: %d (occurs %d times)\n" m_val m_count
